### DISCUSSION TITLE
How do I animate _only a child_ of an element with an animation util class?

### DISCUSSION THREAD

**Discussion by jasonmevans** [tailwindlabs/tailwindcss#13207] [Help] [answered]
## Setup

Given this base configuration,

```js
const baseConfig = {
  content: getContent(),
  presets: [tailwindPreset],
  theme: {
    extend: {
      keyframes: {
        blinking: {
          '0%': {
            opacity: '1',
          },
          '50%': {
            opacity: '1',
          },
          '100%': {
            opacity: '0',
          },
        },
      },
      animation: {
        blinking: 'blinking 0.6s ease-out infinite',
      },
    },
  },
};
```

Which produces this corresponding CSS,

```css
@keyframes blinking {
  0% {
    opacity: 1;
  }
  50% {
    opacity: 1;
  }
  100% {
    opacity: 0;
  }
}

.animate-blinking {
  animation: blinking 0.6s ease-out infinite;
}
```

I can animate an element like,

```html
<div classname="animate-blinking">I'm blinking</div>
```

**This works if I want to animate the element that I have directly applied the `animate-blinking` class to. However, what if I only want to animate a single child of the `animate-blinking` element?** 

## The Problem

I'm attempting to write some CSS to apply an animation to a _child_ of my generated CSS class's target. However, after all that I've attempted, I've found that the associated CSS `keyframes` appear to be emitted conditionally. Meaning, you can only use these keyframes by applying the generated CSS class directly to the element you wish to animate.

In other words, I need to animate the `[data-blink-me="true"]` child of `.animate-blinking`, and not `.animate-blinking` itself, 

```html
<div classname="animate-blinking">
  <div data-blink-me="true">I need to blink, not my parent</div>
</div>
```

However, that does not _appear_ to be possible using TW alone.

## Attempt 1

I've tried simply not using the TW produced `animate-blinking` class, and supplying my own of the form,

```css
.blink-me [data-blink-me="true"] {
  animation: blinking 0.6s ease-out infinite;
}
```

However, the `blinking` keyframes are not available to the `.blink-me` class, `blinking` is greyed out in my Chrome Devtools CSS inspector. [I believe it is due to this](https://tailwindcss.com/docs/functions-and-directives#using-apply-with-per-component-css). 

## Attempt 2

After reading [the article I pasted above](https://tailwindcss.com/docs/functions-and-directives#using-apply-with-per-component-css), I attempted using a TW plugin, which initially seemed promising,

```js
const baseConfig = {
  content: getContent(),
  presets: [tailwindPreset],
  theme: {
    extend: {
      keyframes: {
        blinking: {
          '0%': {
            opacity: '1',
          },
          '50%': {
            opacity: '1',
          },
          '100%': {
            opacity: '0',
          },
        },
      },
      animation: {
        blinking: 'blinking 0.6s ease-out infinite',
      },
    },
  },
  plugins: [
    plugin(function ({ addComponents, theme }) {
      addComponents({
        ".blinking [data-animate-fill='true']": {
          animation: theme('animation.blinking'),
        },
      });
    }),
  ],
};
```

This correctly produces the following CSS,

```css
.blinking [data-blink-me="true"] {
  animation: blinking 0.6s ease-out infinite;
}
```

but again the `blinking` keyframes do not exist for my custom class `.blinking`, and are greyed out in Chrome Devtools. It seems as if TW can't find any classes that reference the keyframes being applied in HTML, and so it does not emit them. If I simply define the keyframes manually it works as expected.

Is this the wrong way to achieve my desired outcome? Should I just KISS?

**Comment by wongjn** [comment, marked as answer]
You could consider:
- Using a `data-*` variant alongside an arbitrary universal child selector arbitrary variant, `[&_*]`:
  ```html
  <div class="data-[blink-me=true]:[&_*]:animate-blinking">
  ```
- Using a single arbitrary variant, `[&_[data-blink-me=true]]:`:
  ```html
  <div class="[&_[data-blink-me=true]]:animate-blinking">
  ```
- Refactoring the previous arbitrary variant solution to a registered variant name in a Tailwind plugin:
  ```js
  plugin(({ addVariant }) => {
    addVariant('foo', '& [data-blink-me=true]');
  });
  ```
  ```html
  <div class="foo:animate-blinking">
  ```
See these solutions working in [this Tailwind Play](https://play.tailwindcss.com/slxPkDhQjb).

**Reply by jasonmevans** [reply]
Thank you @wongjn! These solutions worked quite nicely. It was just difficult to figure out how to get there from a beginner perspective. Where do I send you money, Venmo? lol
